pkg/reconciler/ingress/ingress.go

```diff
@@ -362,39 +357,46 @@ func (c *Controller) dnsRecordFromIngress(ctx context.Context, ingress *networki
 			BlockOwnerDeletion: pointer.Bool(true),
 		},
 	})
-	return record, nil
-}
 
-func (c *Controller) endpointsFromIngress(ctx context.Context, ingress *networkingv1.Ingress) ([]*v1.Endpoint, error) {
-	var endpoints []*v1.Endpoint
+	return c.setEndpointsFromIngress(ctx, ingress, dnsRecord)
+}
 
+func (c *Controller) setEndpointsFromIngress(ctx context.Context, ingress *networkingv1.Ingress, dnsRecord *v1.DNSRecord) error {
 	targets, err := c.targetsFromIngressStatus(ctx, ingress.Status)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
 	hostname := ingress.Annotations[cluster.ANNOTATION_HCG_HOST]
 
-	providerSpecific := v1.ProviderSpecific{
-		{
-			Name:  aws.ProviderSpecificWeight,
-			Value: "100",
-		},
-	}
-
 	for _, target := range targets {
-		endpoint := &v1.Endpoint{
-			DNSName:          hostname,
-			RecordType:       "A",
-			Targets:          []string{target},
-			RecordTTL:        60,
-			ProviderSpecific: providerSpecific,
-			SetIdentifier:    target,
+		var endpoint *v1.Endpoint
+		// Try to find the endpoing that matches this target
+		if dnsRecord.Spec.Endpoints != nil {
+			for _, existingEndpoint := range dnsRecord.Spec.Endpoints {
+				if existingEndpoint.SetIdentifier == target {
+					endpoint = existingEndpoint
+				}
+			}
 		}
-		endpoints = append(endpoints, endpoint)
+
+		// If the endpoint does not exist, add a new one
+		if endpoint == nil {
+			endpoint = &v1.Endpoint{
+				SetIdentifier: target,
+			}
+			dnsRecord.Spec.Endpoints = append(dnsRecord.Spec.Endpoints, endpoint)
+		}
+
+		// Update the endpoint fields
+		endpoint.DNSName = hostname
+		endpoint.RecordType = "A"
+		endpoint.Targets = []string{target}
+		endpoint.RecordTTL = 60
+		endpoint.SetProviderSpecific(aws.ProviderSpecificWeight, "100")
 	}
```

Comment:
Probably personal preference, but I've always found just building a map of existing objects first easier to read, and probably more efficient if that slice was larger.

```suggestion
	}	currentEndpoints := make(map[string]*v1.Endpoint)
	for _, endpoint := range dnsRecord.Spec.Endpoints {
		currentEndpoints[endpoint.SetIdentifier] = endpoint
	}

	for _, target := range targets {
		var endpoint *v1.Endpoint

		endpoint, found := currentEndpoints[target]
		if !found {
			endpoint = &v1.Endpoint{
				SetIdentifier: target,
			}
			dnsRecord.Spec.Endpoints = append(dnsRecord.Spec.Endpoints, endpoint)
		}

		// Update the endpoint fields
		endpoint.DNSName = hostname
		endpoint.RecordType = "A"
		endpoint.Targets = []string{target}
		endpoint.RecordTTL = 60
		endpoint.SetProviderSpecific(aws.ProviderSpecificWeight, "100")
	}
```